File: apps/csharp/Algorithm.Tests/StackQueueTest.cs
using Xunit;
using Algorithm;

namespace Algorithm.Tests;

/// <summary>第4章 スタックとキュー テスト</summary>
public class StackQueueTest
{
    public class FixedStackTest
    {
        private readonly FixedStack<int> _stack = new(64);

        [Fact] public void 初期状態は空() { Assert.True(_stack.IsEmpty()); Assert.False(_stack.IsFull()); }
        [Fact] public void pushしてpeek() { _stack.Push(1); Assert.Equal(1, _stack.Peek()); }
        [Fact] public void pushしてpop() { _stack.Push(1); Assert.Equal(1, _stack.Pop()); }
        [Fact]
        public void 複数pushしてLIFO順にpop()
        { _stack.Push(1); _stack.Push(2); _stack.Push(3); Assert.Equal(3, _stack.Pop()); Assert.Equal(2, _stack.Pop()); Assert.Equal(1, _stack.Pop()); }
        [Fact]
        public void 満杯判定()
        { var s = new FixedStack<int>(3); s.Push(1); s.Push(2); s.Push(3); Assert.True(s.IsFull()); }
        [Fact] public void 空のpopで例外() => Assert.Throws<FixedStack<int>.EmptyException>(() => _stack.Pop());
        [Fact]
        public void 満杯のpushで例外()
        { var s = new FixedStack<int>(2); s.Push(1); s.Push(2); Assert.Throws<FixedStack<int>.FullException>(() => s.Push(3)); }
        [Fact] public void 空のpeekで例外() => Assert.Throws<FixedStack<int>.EmptyException>(() => _stack.Peek());
        [Fact] public void find() { _stack.Push(10); _stack.Push(20); _stack.Push(30); Assert.Equal(1, _stack.Find(20)); }
        [Fact] public void find_見つからない() { _stack.Push(10); Assert.Equal(-1, _stack.Find(99)); }
        [Fact]
        public void count()
        { _stack.Push(5); _stack.Push(5); _stack.Push(10); Assert.Equal(2, _stack.Count(5)); Assert.Equal(1, _stack.Count(10)); Assert.Equal(0, _stack.Count(99)); }
        [Fact] public void contains() { _stack.Push(42); Assert.True(_stack.Contains(42)); Assert.False(_stack.Contains(0)); }
        [Fact] public void clear() { _stack.Push(1); _stack.Push(2); _stack.Clear(); Assert.True(_stack.IsEmpty()); }
        [Fact] public void size() { _stack.Push(1); _stack.Push(2); Assert.Equal(2, _stack.Size()); }
        [Fact] public void capacity() => Assert.Equal(64, _stack.GetCapacity());
    }

    public class FixedQueueTest
    {
        private readonly FixedQueue<int> _queue = new(64);

        [Fact] public void 初期状態は空() { Assert.True(_queue.IsEmpty()); Assert.False(_queue.IsFull()); }
        [Fact] public void enqueしてdeque() { _queue.Enque(1); Assert.Equal(1, _queue.Deque()); }
        [Fact]
        public void 複数enqueしてFIFO順にdeque()
        { _queue.Enque(1); _queue.Enque(2); _queue.Enque(3); Assert.Equal(1, _queue.Deque()); Assert.Equal(2, _queue.Deque()); Assert.Equal(3, _queue.Deque()); }
        [Fact] public void peek() { _queue.Enque(10); _queue.Enque(20); Assert.Equal(10, _queue.Peek()); }
        [Fact]
        public void 満杯判定()
        { var q = new FixedQueue<int>(3); q.Enque(1); q.Enque(2); q.Enque(3); Assert.True(q.IsFull()); }
        [Fact] public void 空のdequeで例外() => Assert.Throws<FixedQueue<int>.EmptyException>(() => _queue.Deque());
        [Fact]
        public void 満杯のenqueで例外()
        { var q = new FixedQueue<int>(2); q.Enque(1); q.Enque(2); Assert.Throws<FixedQueue<int>.FullException>(() => q.Enque(3)); }
        [Fact] public void 空のpeekで例外() => Assert.Throws<FixedQueue<int>.EmptyException>(() => _queue.Peek());
        [Fact] public void find() { _queue.Enque(10); _queue.Enque(20); _queue.Enque(30); Assert.Equal(0, _queue.Find(10)); }
        [Fact] public void find_見つからない() { _queue.Enque(10); Assert.Equal(-1, _queue.Find(99)); }
        [Fact]
        public void count()
        { _queue.Enque(5); _queue.Enque(5); _queue.Enque(10); Assert.Equal(2, _queue.Count(5)); }
        [Fact] public void contains() { _queue.Enque(42); Assert.True(_queue.Contains(42)); Assert.False(_queue.Contains(0)); }
        [Fact] public void clear() { _queue.Enque(1); _queue.Enque(2); _queue.Clear(); Assert.True(_queue.IsEmpty()); }
        [Fact] public void size() { _queue.Enque(1); _queue.Enque(2); Assert.Equal(2, _queue.Size()); }
        [Fact] public void capacity() => Assert.Equal(64, _queue.GetCapacity());
        [Fact]
        public void リングバッファの折り返し()
        {
            var q = new FixedQueue<int>(3);
            q.Enque(1); q.Enque(2); q.Enque(3); q.Deque(); q.Enque(4);
            Assert.Equal(2, q.Deque()); Assert.Equal(3, q.Deque()); Assert.Equal(4, q.Deque());
        }
    }
}

File: apps/java/src/test/java/algorithm/StackQueueTest.java
package algorithm;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class StackQueueTest {

    @Nested
    class FixedStackTest {
        private FixedStack<Integer> stack;

        @BeforeEach
        void setUp() { stack = new FixedStack<>(64); }

        @Test
        void 初期状態は空() {
            assertTrue(stack.isEmpty());
            assertFalse(stack.isFull());
        }

        @Test
        void pushしてpeek() {
            stack.push(1);
            assertEquals(1, stack.peek());
        }

        @Test
        void pushしてpop() {
            stack.push(1);
            assertEquals(1, stack.pop());
        }

        @Test
        void 複数pushしてLIFO順にpop() {
            stack.push(1);
            stack.push(2);
            stack.push(3);
            assertEquals(3, stack.pop());
            assertEquals(2, stack.pop());
            assertEquals(1, stack.pop());
        }

        @Test
        void 満杯判定() {
            FixedStack<Integer> small = new FixedStack<>(3);
            small.push(1); small.push(2); small.push(3);
            assertTrue(small.isFull());
        }

        @Test
        void 空のpopで例外() {
            assertThrows(FixedStack.EmptyException.class, () -> stack.pop());
        }

        @Test
        void 満杯のpushで例外() {
            FixedStack<Integer> small = new FixedStack<>(2);
            small.push(1); small.push(2);
            assertThrows(FixedStack.FullException.class, () -> small.push(3));
        }

        @Test
        void 空のpeekで例外() {
            assertThrows(FixedStack.EmptyException.class, () -> stack.peek());
        }

        @Test
        void find() {
            stack.push(10); stack.push(20); stack.push(30);
            assertEquals(1, stack.find(20));
        }

        @Test
        void find_見つからない() {
            stack.push(10);
            assertEquals(-1, stack.find(99));
        }

        @Test
        void count() {
            stack.push(5); stack.push(5); stack.push(10);
            assertEquals(2, stack.count(5));
            assertEquals(1, stack.count(10));
            assertEquals(0, stack.count(99));
        }

        @Test
        void contains() {
            stack.push(42);
            assertTrue(stack.contains(42));
            assertFalse(stack.contains(0));
        }

        @Test
        void clear() {
            stack.push(1); stack.push(2);
            stack.clear();
            assertTrue(stack.isEmpty());
        }

        @Test
        void size() {
            stack.push(1); stack.push(2);
            assertEquals(2, stack.size());
        }

        @Test
        void capacity() {
            assertEquals(64, stack.getCapacity());
        }
    }

    @Nested
    class FixedQueueTest {
        private FixedQueue<Integer> queue;

        @BeforeEach
        void setUp() { queue = new FixedQueue<>(64); }

        @Test
        void 初期状態は空() {
            assertTrue(queue.isEmpty());
            assertFalse(queue.isFull());
        }

        @Test
        void enqueしてdeque() {
            queue.enque(1);
            assertEquals(1, queue.deque());
        }

        @Test
        void 複数enqueしてFIFO順にdeque() {
            queue.enque(1); queue.enque(2); queue.enque(3);
            assertEquals(1, queue.deque());
            assertEquals(2, queue.deque());
            assertEquals(3, queue.deque());
        }

        @Test
        void peek() {
            queue.enque(10); queue.enque(20);
            assertEquals(10, queue.peek());
        }

        @Test
        void 満杯判定() {
            FixedQueue<Integer> small = new FixedQueue<>(3);
            small.enque(1); small.enque(2); small.enque(3);
            assertTrue(small.isFull());
        }

        @Test
        void 空のdequeで例外() {
            assertThrows(FixedQueue.EmptyException.class, () -> queue.deque());
        }

        @Test
        void 満杯のenqueで例外() {
            FixedQueue<Integer> small = new FixedQueue<>(2);
            small.enque(1); small.enque(2);
            assertThrows(FixedQueue.FullException.class, () -> small.enque(3));
        }

        @Test
        void 空のpeekで例外() {
            assertThrows(FixedQueue.EmptyException.class, () -> queue.peek());
        }

        @Test
        void find() {
            queue.enque(10); queue.enque(20); queue.enque(30);
            assertEquals(0, queue.find(10));
        }

        @Test
        void find_見つからない() {
            queue.enque(10);
            assertEquals(-1, queue.find(99));
        }

        @Test
        void count() {
            queue.enque(5); queue.enque(5); queue.enque(10);
            assertEquals(2, queue.count(5));
        }

        @Test
        void contains() {
            queue.enque(42);
            assertTrue(queue.contains(42));
            assertFalse(queue.contains(0));
        }

        @Test
        void clear() {
            queue.enque(1); queue.enque(2);
            queue.clear();
            assertTrue(queue.isEmpty());
        }

        @Test
        void size() {
            queue.enque(1); queue.enque(2);
            assertEquals(2, queue.size());
        }

        @Test
        void capacity() {
            assertEquals(64, queue.getCapacity());
        }

        @Test
        void リングバッファの折り返し() {
            FixedQueue<Integer> small = new FixedQueue<>(3);
            small.enque(1); small.enque(2); small.enque(3);
            small.deque(); // 1 を取り出す
            small.enque(4); // 空き位置に追加
            assertEquals(2, small.deque());
            assertEquals(3, small.deque());
            assertEquals(4, small.deque());
        }
    }
}
